osu.Game/Database/ArchiveModelManager.cs

```diff
@@ -130,56 +132,50 @@ protected ArchiveModelManager(Storage storage, IDatabaseContextFactory contextFa
         /// This will post notifications tracking progress.
         /// </summary>
         /// <param name="paths">One or more archive locations on disk.</param>
-        public void Import(params string[] paths)
+        public async Task Import(params string[] paths)
         {
             var notification = new ProgressNotification { State = ProgressNotificationState.Active };
 
             PostNotification?.Invoke(notification);
-            Import(notification, paths);
+
+            await Import(notification, paths);
         }
 
-        protected void Import(ProgressNotification notification, params string[] paths)
+        protected async Task Import(ProgressNotification notification, params string[] paths)
         {
             notification.Progress = 0;
             notification.Text = "Import is initialising...";
 
             var term = $"{typeof(TModel).Name.Replace("Info", "").ToLower()}";
 
-            List<TModel> imported = new List<TModel>();
-
             int current = 0;
 
-            foreach (string path in paths)
+            var imported = new List<TModel>();
+
+            await Task.WhenAll(paths.Select(path => Import(path, notification.CancellationToken).ContinueWith(t =>
```

Comment:
Internally, async & await uses continuations. Seeing the usage of `ContinueWith` here, I feel this should have been done with asyncs and awaits.

Reply:
How would you propose this one is written? The ContinueWith is being applied to each task.

Reply:
What about this? Though, I am not sure this is 100% the same:
```csharp
var tasksTemp = paths.Select(doSomethingAsync);

async Task doSomethingAsync(string path)
{
    try
    {
        TModel result = await Import(path, notification.CancellationToken);

        lock (notification)
        {
            current++;

            notification.Text = $"Imported {current} of {paths.Length} {term}s";
            notification.Progress = (float)current / paths.Length;
        }

        lock (imported)
            imported.Add(result);
    }
    catch (OperationCanceledException)
    {
        // do nothing?
    }
    catch (Exception e)
    {
        Logger.Error(e, $@"Could not import ({Path.GetFileName(path)})");
    }
}

await Task.WhenAll(tasksTemp);
```

Reply:
(Of course, with a lambda instead of a local method if you prefer)

Reply:
looks around right. will give it a shot and see if it reads better.

Reply:
Another thing that came to my mind - and I want to note - is that in the original control flow, the could would say "Imported xyz" and only after that it checks for an exception on the previous task.
(By the way I am not sure that Continuations run when the previous one throwed an exception)